javaagent-api/src/main/jflex/SqlSanitizer.flex

```diff
@@ -0,0 +1,378 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.api.db.sanitizer;
+
+import io.opentelemetry.javaagent.instrumentation.api.db.SqlStatementInfo;
+
+%%
+
+%public
+%final
+%class SqlSanitizer
+%apiprivate
+%int
+%buffer 2048
+
+%unicode
+%ignorecase
+
+COMMA             = ","
+OPEN_PAREN        = "("
+CLOSE_PAREN       = ")"
+OPEN_COMMENT      = "/*"
+CLOSE_COMMENT     = "*/"
+IDENTIFIER        = ([:letter:] | "_") ([:letter:] | [0-9] | [_.])*
+BASIC_NUM         = [.+-]* [0-9] ([0-9] | [eE.+-])*
+HEX_NUM           = "0x" ([a-f] | [A-F] | [0-9])+
+QUOTED_STR        = "'" ("''" | [^'])* "'"
+DOUBLE_QUOTED_STR = "\"" ("\"\"" | [^\"])* "\""
+DOLLAR_QUOTED_STR = "$$" [^$]* "$$"
+WHITESPACE        = [ \t\r\n]+
+
+%{
+  public static SqlStatementInfo sanitize(String statement) {
+    if (statement == null) {
+        return new SqlStatementInfo(null, null, null);
+    }
+    SqlSanitizer sanitizer = new SqlSanitizer(new java.io.StringReader(statement));
```

Comment:
Yeah, that would be awesome... and probably would improve the performance even more, since char[] copying seems to take a lot of time in Jason's perf test. Unfortunately jflex does not allow this degree of customisation.

The generated code is pretty much unreadable.
I think that maybe we could use a semi-manual approach: keep a flex grammar file + e.g. patch file with customizations applied to the generated code; whenever somebody has to change something they would have to generate+apply changes by themselves. And fix the patch when it breaks 😅